src/Shipments/PriceCalculator.php

```diff
@@ -3,155 +3,134 @@
 namespace MyParcelCom\ApiSdk\Shipments;
 
 use MyParcelCom\ApiSdk\Exceptions\CalculationException;
-use MyParcelCom\ApiSdk\Resources\Interfaces\ServiceContractInterface;
-use MyParcelCom\ApiSdk\Resources\Interfaces\ServiceGroupInterface;
+use MyParcelCom\ApiSdk\Exceptions\InvalidResourceException;
+use MyParcelCom\ApiSdk\Resources\Interfaces\ServiceRateInterface;
 use MyParcelCom\ApiSdk\Resources\Interfaces\ShipmentInterface;
 
 class PriceCalculator
 {
     /**
-     * Calculate the total price of given shipment. Optionally a contract can be
-     * supplied for the price calculations. If no contract is given, the
-     * contract on the shipment is used.
+     * Calculate the total price of given shipment based off its contract, service and weight.
      *
-     * @param ShipmentInterface             $shipment
-     * @param ServiceContractInterface|null $contract
+     * @param ShipmentInterface         $shipment
+     * @param ServiceRateInterface|null $serviceRate
      * @return int
      */
-    public function calculate(ShipmentInterface $shipment, ServiceContractInterface $contract = null)
+    public function calculate(ShipmentInterface $shipment, ServiceRateInterface $serviceRate = null)
     {
-        // TODO: Fix!
-        if ($contract === null) {
-            $contract = $shipment->getServiceContract();
+        if ($shipment->getPhysicalProperties() === null
+            || $shipment->getPhysicalProperties()->getWeight() === null
+            || $shipment->getPhysicalProperties()->getWeight() < 0) {
+            throw new InvalidResourceException(
+                'Cannot calculate shipment price without a valid shipment weight.'
+            );
+        }
+
+        if ($serviceRate === null) {
+            $serviceRate = $this->determineServiceRateForShipment($shipment);
         }
 
-        if ($contract === null) {
+        if ($shipment->getPhysicalProperties()->getWeight() < $serviceRate->getWeightMin()
+            || $shipment->getPhysicalProperties()->getWeight() > $serviceRate->getWeightMax()) {
```

Comment:
You're not using the `step_size` and `step_price` in your calculations.

Reply:
That is correct. We currently do not have any services that support this. I will discuss with Lars and Dolores if we still want this, and if we do, make a proposal on how to do it in the API. It is currently implemented a bit strangely, so I decided to leave it out here. The API also doesn't take it into account anymore when returning service rates.